Mathlib/Algebra/Category/AlgebraCat/Basic.lean

```diff
@@ -46,22 +47,90 @@ instance : CoeSort (AlgebraCat R) (Type v) :=
 
 attribute [coe] AlgebraCat.carrier
 
+/-- The object in the category of R-algebras associated to a type equipped with the appropriate
+typeclasses. -/
+def of (X : Type v) [Ring X] [Algebra R X] : AlgebraCat.{v} R :=
+  ⟨X⟩
+
+@[simp]
+lemma coe_of (X : Type v) [Ring X] [Algebra R X] : (of R X : Type v) = X :=
+  rfl
+
+variable {R} in
+/-- The type of morphisms in `AlgebraCat R`. -/
+@[ext]
+structure Hom (A B : AlgebraCat.{v} R) where
+  /-- The underlying algebra map. -/
+  algHom : A →ₐ[R] B
+
 instance : Category (AlgebraCat.{v} R) where
-  Hom A B := A →ₐ[R] B
-  id A := AlgHom.id R A
-  comp f g := g.comp f
+  Hom A B := Hom A B
+  id A := ⟨AlgHom.id R A⟩
+  comp f g := ⟨g.algHom.comp f.algHom⟩
```

Comment:
Changed, I _think_ I found all of them, but if not, please say so.

Reply:
I don't necessarily think it will help with performance much. The main thing is that Lean can choke on eta reduction when it is deep enough in a term so it is less problematic theoretically to avoid the extra eta expansions. But perhaps. 

I think the main performance penalty is a lot of failed simp unification attempts.

Reply:
The main slowdown seems to come from synthesizing `MulHomClass` etc. around L103 of `AlgebraCat.Limits`. This is because we are no longer allowed to use `AlgHom.map_mul` as a shortcut (using these roughly gets build time back to the `FunLike` version).

Reply:
Why was the suggestion reverted?

Reply:
Because it caused a slowdown of another +4⬝10⁹ instructions and because I wanted to get numbers for the `map_mul` -> `AlgHom.map_mul` change independently of your suggestion.

Reply:
Happy to revert the revert if you think it is better, but it seems strange to me that performance got worse.

Reply:
I think it is more a symptom of problems with the current design - like the too generic lemmas - than anything else. Writing out eta expanded terms when unnecessary is inadvisable. I would prefer it be avoided here.

Reply:
Okay, re-applied the commit.